### PR TITLE
[JBPM-9923] OptimisticLockRetryInterceptor retry causes data update instead of insertion when persisting custom entity using Pluggable Variable Persistence

### PR DESCRIPTION
**Pull request by elguardian**
Jira: https://issues.redhat.com/browse/JBPM-9923

### CODE REVIEW

**Comment by elguardian** [comment]
Jenkins run fdb

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by gmunozfe** [comment]
jenkins retest this

File: drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java [hunk 1]
```diff
@@ -22,14 +22,18 @@
...
 import java.util.HashSet;
+import java.util.List;
 import java.util.Set;
 
 import javax.persistence.EntityManager;
 import javax.persistence.EntityManagerFactory;
 import javax.persistence.Persistence;
 import javax.persistence.metamodel.EntityType;
 import javax.persistence.metamodel.Metamodel;
+import javax.persistence.metamodel.Type;
```
Comment (gmunozfe) [inline drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java]: This is not used

File: drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java [hunk 2]
```diff
@@ -207,6 +211,33 @@ public void onStart(TransactionManager txm) {
...
+            // this is pretty much of a hack but for avoiding issues when rolling back we need to set to null 
+            // the primary key of the entities (simple types)
+            List<Object> entities = em.getEntities();
+            entities.stream().forEach(entity -> {
+                try {
+                    Metamodel metamodel = emf.getMetamodel();
+                    EntityType<? extends Object> jpaEntity = metamodel.entity(entity.getClass());
+                    String idFieldName = jpaEntity.getId(jpaEntity.getIdType().getJavaType()).getName();
+	            Field field = null;
+		    while(clazz != null) {
```
Comment (gmunozfe) [inline drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java]: Check indentation and add a blank before the parenthesis.

File: drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java [hunk 3]
```diff
@@ -207,6 +211,33 @@ public void onStart(TransactionManager txm) {
...
+			} catch (Throwable th) {
+			    clazz = clazz.getSuperclass();
+			}
+		    }
+		    if (field != null) {
+                        field.setAccessible(true);
+                        field.set(entity, null);
+		    }
+                } catch(Exception e) {
+                    log.debug("There is a problem trying to rollaback id object {}", entity, e);
```
Comment (gmunozfe) [inline drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java]: typo "rollback"

File: drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java (hunk 2 above)
Comment (gmunozfe) [inline drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java]: Compilation error, clazz not defined

**Review by gmunozfe — CHANGES_REQUESTED** [review]

File: drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java (hunk 2 above)
Comment (elguardian) [reply drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java]: done

File: drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java (hunk 3 above)
Comment (elguardian) [reply drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java]: done

File: drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java (hunk 2 above)
Comment (elguardian) [reply drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java]: done

**Comment by elguardian** [comment]
Jenkins retest this please

**Review by gmunozfe — APPROVED** [review]
Looks good to me, though I couldn't find a way to test it. Fantastic work @elguardian !

**Pull request merged**